### PR TITLE
Add refactor-breakpointmods rule

### LINKED ISSUE #70: Add refactor-breakpointMods
https://github.com/patternfly/patternfly-react/pull/4310

Before:
```js
breakpointMods={[
  { modifier: 'visible' },
  { modifier: 'hidden', breakpoint: 'lg' }
]}
```

After:
```js
visibility={{
  default: 'visible',
  lg: 'hidden'
}}
```

## Breaking changes
- **Flex:**
  - Removed `breakpointMods` prop from `Flex` in favor of `spacer`, `spaceItems`, `grow`, `shrink`, `flex`, `direction`, `alignItems`, `alignContent`, `alignSelf`, `align`, `justifyContent`, `display`, `fullWidth` and `flexWrap`
  - Removed `breakpointMods` prop from `FlexItem` in favor of `spacer`, `grow`, `shrink`, `flex`, `alignSelf`, `align`, and `fullWidth`
- **Page:**
  - Removed prop `hasNoPadding` in favor of `padding={{ default: 'noPadding' }}` to prevent conflicts when setting both `hasNoPadding` and `padding={{ default: 'padding' | 'noPadding' }}`
- **Toolbar:**
  - Removed `breakpointMods` prop from `ToolbarItem` in favor of `visiblity`, `alignment`, and `spacer`
  - Removed `breakpointMods` prop from `ToolbarToggleGroup` in favor of `visiblity`, `alignment`, `spacer`, and `spaceItems`
  - Removed `breakpoint` prop from `ToolbarToggleGroup` in favor of `show`

### PR DESCRIPTION
**Pull request by evwilkin**
Closes #70 

This PR:
- removes & replaces `breakpointMods` prop from `Flex, FlexItem, ToolbarItem,, and ToolbarToggleGroup` components;
- removes & replaces `hasNoPadding` prop from `PageSection` component;
- removes & replaces `breakpoint` prop from `ToolbarToggleGroup` component.

See #70 for full details of changes.

### CODE REVIEW

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+          {modifier: 'spacer-none'},
+          {modifier: 'row', breakpoint: 'sm'}
+        ]}>
+          <FlexItem breakpointMods={[
+            {breakpoint: 'xl', modifier: FlexModifiers.column},
+            {modifier: FlexModifiers["align-self-stretch"], breakpoint: 'md'},
+            {modifier: FlexModifiers.row}
+          ]}></FlexItem>
+        </Flex>`,
+      output: `import { Flex, FlexItem, FlexModifiers } from '@patternfly/react-core';
```
Comment (redallen) [inline packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:39]: We also need to remove the imports `'FlexModifiers', 'FlexBreakpoints', /\w+BreakpointMod/`

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  align: ['alignLeft', 'alignRight'],
+  justifyContent: ['justifyContentFlexStart', 'justifyContentFlexEnd', 'justifyContentCenter', 'justifyContentSpaceBetween', 'justifyContentSpaceAround', 'justifyContentSpaceEvenly'],
+  display: ['inlineFlex'],
+  fullWidth: ['fullWidth'],
+  flexWrap: ['wrap', 'wrapReverse', 'nowrap']
+};
+const errorMessages = {
+  Flex: 'Removed breakpointMods prop from Flex in favor of spacer, spaceItems, grow, shrink, flex, direction, alignItems, alignContent, alignSelf, align, justifyContent, display, fullWidth and flexWrap',
+  FlexItem: 'Removed breakpointMods prop from FlexItem in favor of spacer, grow, shrink, flex, alignSelf, align, and fullWidth',
+  ToolbarItem: 'Removed breakpointMods prop from ToolbarItem in favor of visiblity, alignment, and spacer',
```
Comment (redallen) [inline packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:34]: Need PageSection in here. Sorry for my incorrect note!

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+          } else {
+            // ToolbarToggleGroup
+            if (importName === 'ToolbarToggleGroup') {
+              const attribute = node.attributes.find(attr => attr.name && attr.name.name === 'breakpoint');
+              if (attribute) {
+                const breakpointVal = attribute.value.value;
+                const newAttr = `show={{ ${breakpointVal}: 'show' }}`
+                context.report({
+                  node,
+                  message: `hasNoPadding prop on Page component removed in favor of padding={{ default: 'noPadding' }}`,
```
Comment (jschuler) [inline packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:72]: this message needs to be updated

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                }
+              });
+            }
+          } else {
+            // ToolbarToggleGroup
+            if (importName === 'ToolbarToggleGroup') {
+              const attribute = node.attributes.find(attr => attr.name && attr.name.name === 'breakpoint');
+              if (attribute) {
+                const breakpointVal = attribute.value.value;
+                const newAttr = `show={{ ${breakpointVal}: 'show' }}`
```
Comment (jschuler) [inline packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:69]: is this logically correct? Looking at the old code, if `breakpoint="md"` it seems that it was supposed to collapse/hide at that breakpoint.

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js (hunk 3 above)
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:72]: Updated error message

[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:72]: Updated error message

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            const attribute = node.attributes.find(attr => attr.name && attr.name.name === 'breakpointMods');
+            if (
+              attribute &&
+              attribute.value.type === 'JSXExpressionContainer' &&
+              attribute.value.expression &&
+              attribute.value.expression.type === 'ArrayExpression'
+            ) {
+              const breakpointModsArr = attribute.value.expression.elements;
+              const newModifier = {};
+              breakpointModsArr.map(breakpointModObj => {
```
[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:89]: Updated .map to .forEach

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
+const ruleTester = require('./ruletester');
+const rule = require('../../lib/rules/refactor-breakpointmods');
+
+ruleTester.run("refactor-breakpointmods", rule, {
+  valid: [
+    {
+      code: `<Flex alignSelf={{default: "alignSelfStretch"}}>
```
[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:7]: Added imports

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        <FlexItem alignSelf={{default: "alignSelfStretch"}}></FlexItem>
+      </Flex>`
+    },
+    {
+      code: `<PageSection padding={{ default: 'noPadding' }}></PageSection>`
```
[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:12]: Added import

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    },
+    {
+      code: `<ToolbarItem visibility={{
```
[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:15]: Added import statement

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        default: 'visible',
+        lg: 'hidden'
+      }}></ToolbarItem>`
+    },
+    {
+      code: `<ToolbarToggleGroup spacer={{default: "spacerNone"}} show={{xl: 'show'}}></ToolbarToggleGroup>`
```
[the remark this answers is not in the document]
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:21]: Added import statement

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js (hunk 2 above)
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:34]: No update needed - PageSection error message is hardcoded inline as there are no dynamic use cases when reaching this error

File: packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js (hunk 1 above)
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/test/rules/refactor-breakpointmods.js:39]: Updated `remove-unused-imports` rule to remove `FlexModifiers, FlexBreakpoints, FlexBreakpointMod, FlexItemBreakpointMod`.  There is still a `formatBreakpointMods` method used in react-core that I did not want to mistakenly remove using a broader regex match.

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js (hunk 4 above)
Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:69]: @jschuler thanks for calling this out - I believe the description of the previous prop was a bit confusing, it listed the `breakpoint` prop as `The breakpoint at which the toggle group is collapsed down` but is actually used as the breakpoint at which the toggle group is expanded (anything below it is collapsed) as seen in the prior usage:
`styles.modifiers[`showOn${capitalize(breakpoint)}` as 'showOnMd' | 'showOnLg' | 'showOnXl'],`

I believe the usage is correct as is in the PR, and the verbiage of `show` is a little more accurate as it's the breakpoint at which the ToolbarToggleGroup is shown in place of just the toggle icon.

Comment (evwilkin) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:69]: @redallen does this match with your understanding when implementing the new API?

Comment (redallen) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:69]: Yep! That's correct @evwilkin . I can see why the confusion arose...

**Comment by redallen** [comment]
@evwilkin Merge conflict in the index file

File: packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js (hunk 4 above)
Comment (jschuler) [reply packages/eslint-plugin-pf-codemods/lib/rules/refactor-breakpointmods.js:69]: Thanks for checking @evwilkin !

**Pull request merged**